src/controller/python/chip/ble/LinuxImpl.cpp

```diff
@@ -11,41 +12,125 @@ using namespace chip::DeviceLayer::Internal;
 {
-    return static_cast<chip::DeviceLayer::Internal::AdapterIterator *>(adapter)->GetAddress();
+    return static_cast<AdapterIterator *>(adapterIterator)->GetAddress();
 }
 
-extern "C" const char * pychip_ble_adapter_list_get_alias(void * adapter)
+extern "C" const char * pychip_ble_adapter_list_get_alias(void * adapterIterator)
 {
-    return static_cast<chip::DeviceLayer::Internal::AdapterIterator *>(adapter)->GetAlias();
+    return static_cast<AdapterIterator *>(adapterIterator)->GetAlias();
 }
 
-extern "C" const char * pychip_ble_adapter_list_get_name(void * adapter)
+extern "C" const char * pychip_ble_adapter_list_get_name(void * adapterIterator)
 {
-    return static_cast<chip::DeviceLayer::Internal::AdapterIterator *>(adapter)->GetName();
+    return static_cast<AdapterIterator *>(adapterIterator)->GetName();
 }
 
-extern "C" bool pychip_ble_adapter_list_is_powered(void * adapter)
+extern "C" bool pychip_ble_adapter_list_is_powered(void * adapterIterator)
 {
-    return static_cast<chip::DeviceLayer::Internal::AdapterIterator *>(adapter)->IsPowered();
+    return static_cast<AdapterIterator *>(adapterIterator)->IsPowered();
+}
+
+extern "C" void * pychip_ble_adapter_list_get_raw_adapter(void * adapterIterator)
+{
+    return static_cast<AdapterIterator *>(adapterIterator)->GetAdapter();
+}
+
+/////////// CHIP Device scanner implementation //////////
+
+namespace {
+
+// To avoid pythoon compatibility issues on inc/dec references,
+// code assumes an abstract type and leaves it up to python to keep track of
+// reference counts
+struct PyObject;
+
+class ScannerDelegateImpl : public ChipDeviceScannerDelegate
+{
+public:
+    using DeviceScannedCallback = void (*)(PyObject * context, const char * address, uint16_t discriminator, uint16_t vendorId,
+                                           uint16_t productId);
+    using ScanCompleteCallback  = void (*)(PyObject * context);
+
+    ScannerDelegateImpl(PyObject * context, DeviceScannedCallback scanCallback, ScanCompleteCallback completeCallback) :
+        mContext(context), mScanCallback(scanCallback), mCompleteCallback(completeCallback)
+    {}
+
+    ~ScannerDelegateImpl()
+    {
+        mScanCallback     = nullptr;
+        mCompleteCallback = nullptr;
+    }
+
+    void SetScanner(std::unique_ptr<ChipDeviceScanner> scanner) { mScanner = std::move(scanner); }
+
+    void OnDeviceScanned(BluezDevice1 * device, const chip::Ble::ChipBLEDeviceIdentificationInfo & info) override
+    {
+        if (mScanCallback)
+        {
+            mScanCallback(mContext, bluez_device1_get_address(device), info.GetDeviceDiscriminator(), info.GetProductId(),
+                          info.GetVendorId());
+        }
+    }
+
+    void OnScanComplete() override
+    {
+        if (mCompleteCallback)
+        {
+            mCompleteCallback(mContext);
+        }
+
+        delete this;
+    }
+
+private:
+    std::unique_ptr<ChipDeviceScanner> mScanner;
```

Comment:
nit: const

Reply:
Cannot do this because scanner is moved post object creation and I have a SetScanner method. 

This is a side-effect of creation order: delegate needs to access the scanner, but scanner needs a delegate.